pandas/io/excel/_calamitereader.py

```diff
@@ -0,0 +1,69 @@
+from __future__ import annotations
+
+from io import (
+    BufferedReader,
+    BytesIO,
+)
+from pathlib import PurePath
+from tempfile import NamedTemporaryFile
+from typing import Any
+
+from python_calamine import (
+    get_sheet_data,
+    get_sheet_names,
+)
+
+from pandas.io.excel._base import (
+    BaseExcelReader,
+    inspect_excel_format,
+)
+
+
+class __calamine__:
+    pass
+
+
+class CalamineExcelReader(BaseExcelReader):
+    book: str
+    _sheet_names: list[str] | None = None
+
+    @property
+    def _workbook_class(self) -> type[__calamine__]:
+        return __calamine__
+
+    def load_workbook(
+        self, filepath_or_buffer: str | PurePath | BufferedReader | BytesIO
+    ) -> str:
+        if isinstance(filepath_or_buffer, BufferedReader):
+            filepath_or_buffer = filepath_or_buffer.name
+
+        elif isinstance(filepath_or_buffer, BytesIO):
+            ext = inspect_excel_format(filepath_or_buffer)
+            with NamedTemporaryFile(suffix=f".{ext}", delete=False) as tmp_file:
+                tmp_file.write(filepath_or_buffer.getvalue())
+                filepath_or_buffer = tmp_file.name
+
+        elif isinstance(filepath_or_buffer, PurePath):
+            filepath_or_buffer = filepath_or_buffer.as_posix()
+
+        assert isinstance(filepath_or_buffer, str)
+
+        self._sheet_names = get_sheet_names(filepath_or_buffer)
+        return filepath_or_buffer
+
+    @property
+    def sheet_names(self) -> list[str]:
+        if self._sheet_names is None:
+            self._sheet_names = get_sheet_names(self.book)
+        return self._sheet_names
+
+    def get_sheet_by_name(self, name: str) -> int:
+        self.raise_if_bad_sheet_by_name(name)
+        return self.sheet_names.index(name)
+
+    def get_sheet_by_index(self, index: int) -> int:
+        self.raise_if_bad_sheet_by_index(index)
+        return index
```

Comment:
I presume there is more being added here?

Reply:
This was ported over from the implementation in the library's repo [here](https://github.com/dimastbk/python-calamine/blob/master/python/python_calamine/pandas.py), but obvs wasn't implemented by the maintainer.

I'm unsure if this is something the maintainer of the library will add in the future but I can remove this for now.

Reply:
`python-calamine` isn't returning sheet from calamine (you can use [fastexcel](https://github.com/ToucanToco/fastexcel) if you need sheet). This functions are used only in https://github.com/pandas-dev/pandas/blob/1f836f16b15adc0838afe634df5bda3977b3ad10/pandas/io/excel/_base.py#L733_L736 and working in this case.

Reply:
Fixed.